enhancements/rhcos/rhcos-cloud-marketplace-images.md

```diff
@@ -0,0 +1,330 @@
+---
+title: marketplace-images
+authors:
+  - "@patrickdillon"
+reviewers: # Include a comment about what domain expertise a reviewer is expected to bring and what area of the enhancement you expect them to focus on. For example: - "@networkguru, for networking aspects, please look at IP bootstrapping aspect"
+  - TBD
+approvers: # A single approver is preferred, the role of the approver is to raise important questions, help ensure the enhancement receives reviews from all applicable areas/SMEs, and determine when consensus is achieved such that the EP can move forward to implementation.  Having multiple approvers makes it difficult to determine who is responsible for the actual approval.
+  - "@sdodson"
+api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
+  - "None"
+creation-date: 2024-07-12
+last-updated: 2024-07-12
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - "SPSTRAT-271"
+see-also:
+replaces:
+superseded-by:
+---
+
+# RHCOS Cloud-Marketplace Images
+
+## Summary
+
+This enhancement proposes a process to upload/publish
+cloud (AWS, Azure, & GCP) marketplace images and include
+the image details in the RHCOS stream.
+
+## Motivation
+
+Publishing marketplace versions in the RHCOS stream will
+improve user experience and relieve tech debt:
+
+* Marketplace images published as part of Project Stratosphere are incomplete, requiring users to use
+boot images from previous releases, which slows
+down machine provisioning.
```

Comment:
I think this is a bit oversold, sometimes even in the z-stream that updates boot images the boot image doesn't match the rpm-ostree so there's still a pivot because there's a timing misalignment. I wouldn't expect this to significantly impact provisioning times for most clusters.

Reply:
I read secondhand in a bug report that because we are publishing old y-streams (e.g. booting a 4.16 cluster with 4.13 images) it requires multiple reboots/pivots/or something else that takes significant amount of time. I can dig up the bug.

Reply:
For HCP, node pools at a newer version would require copying the associated images for each change in a z-stream, so it might be a bit oversold for OCP, but for HCP there's a price to pay every net new nodepool creation at a newer version.

Reply:
An upgraded cluster, at least standalone does boot the old image apply an rpm-ostree rebase then reboot. If the node is part of a custom MCP there's an additional reboot once the node gets labeled properly. However, unless we ship boot images for every z-stream AND ensure that the boot image matches the rpm-ostree which I think would be a significant challenge to coordinate, that reboot will always exist and the amount of time consumed going from a 4.1 to 4.17 boot image is not significantly more than 4.17.0 to 4.17.1, most of the time is consumed with starting everything up, fetching the container image from the network, rebooting.

A quick read through the nodepool controller in hypershift seems to indicate that if the image is set when creating a nodepool it will use that image, if it isn't the AWS code uses the boot image from the release image and for Azure it seems to use a [static image.](https://github.com/openshift/hypershift/blob/8b880bdd20d12de5f0e61dc048bc0b5d0f3ea015/hypershift-operator/controllers/nodepool/azure.go#L105-L110)

Reply:
I think the AWS link might be wrong 😄 Maybe it should be [this](https://github.com/openshift/hypershift/blob/21759c4244d463662bce675046b9d72488ca3f21/hypershift-operator/controllers/nodepool/nodepool_controller.go#L2492)?

For Azure, it's based off of a VHD Image ID specified in the NodePool CR atm. We have a PR open that adds support to get the RHCOS image from an Azure Marketplace.

Reply:
@bryan-cox When someone updates a pool from one version do another do these get updated?

Reply:
For ROSA we [pick whatever bootimage comes within the payload you are upgrading towards](https://github.com/openshift/hypershift/blob/8b880bdd20d12de5f0e61dc048bc0b5d0f3ea015/hypershift-operator/controllers/nodepool/nodepool_controller.go#L2492-L2505) and recreate machines

Reply:
Ok, so without checking to see if bootimage == osImageURL at most this is true about 10% of the time, 6 boot images have shipped for 61 z-streams in 4.12 and those are disproportionately skewed to earlier z-streams.
- 4.12.0
- 4.12.2
- 4.12.3
- 4.12.10
- 4.12.17
- 4.12.30